### PR TITLE
Support for Actor timers, and fix bug in FSM, #15733

### PR DESCRIPTION
**Pull request by patriknw**
* backport of the timers from Akka Typed, #16742
* also fixed a small bug in FSM timers, which could result in that
  a timer from a previous incarnation was let through to new
  incarnation after restart
* no more need for the complicated "how to" section in docs of
  how to schedule periodic messages

Refs #15733

### CODE REVIEW

File: project/MiMa.scala [hunk 1]
```diff
@@ -1159,6 +1160,14 @@ object MiMa extends AutoPlugin {
...
         ProblemFilters.exclude[ReversedMissingMethodProblem]("akka.remote.WireFormats#DeployDataOrBuilder.getScopeManifest")
       )
+// FIXME      
+//      ,
+//      "2.4.18" -> Seq(
+//        // #15733 Timers
+//        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.persistence.fsm.PersistentFSM#Timer.apply"),
+//        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.persistence.fsm.PersistentFSM#Timer.copy"),
+//        ProblemFilters.exclude[DirectMissingMethodProblem]("akka.persistence.fsm.PersistentFSM#Timer.this")
+//      )
```
Comment (patriknw) [inline project/MiMa.scala]: @2m I can't get this to work. I have thrown those lines at several places but nothing sticks (I don't understand).

File: akka-actor-tests/src/test/scala/akka/actor/TimerSpec.scala [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+      val probe = TestProbe()
+      val ref = system.actorOf(target(probe.ref, dilatedInterval, repeat = true))
+      ref ! Throw(new Exc)
+      probe.expectMsg(GotPreRestart(false))
+
+      ref ! End
+      probe.expectMsg(GotPostStop(false))
+    }
+
+    "discard timers from old incarnation after restart, alt 1" taggedAs TimingTest in {
```
Comment (patriknw) [inline akka-actor-tests/src/test/scala/akka/actor/TimerSpec.scala]: this is the reproducer for the FSM bug

File: akka-actor/src/main/scala/akka/actor/Actor.scala [hunk 3]
```diff
@@ -272,6 +272,7 @@ class ActorInterruptedException private[akka] (cause: Throwable) extends AkkaExc
  */
 @SerialVersionUID(1L)
 final case class UnhandledMessage(@BeanProperty message: Any, @BeanProperty sender: ActorRef, @BeanProperty recipient: ActorRef)
+  extends NoSerializationVerificationNeeded
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/Actor.scala]: I noticed that if a message that is not serializable is published as unhandled then the unhandled message is reported as not serializable and hiding the root cause

**Comment by 2m** [comment]
Added a commit that shuffles filters around and makes MiMa happy now. There were two things that needed to be fixed:

* fitlers for `akka.actor.FSM#Timer.*` were missing
* after adding 2.4.18 to the list, some of the BC issues reappeared because filters for those were in the 2.4.17 filter list. These were moved to `bcIssuesBetween24and25`.

File: akka-actor/src/main/scala/akka/actor/Timers.scala [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+   * Java API: Factory method for [[TimerScheduler]]. `TimerScheduler` is
+   * typically with `AbstractActorWithTimers` but in some case the single inheritance
+   * slot is already taken, by for example `AbstractPersistentActor`, and then the
+   * timers can be created and managed by composition instead. Then the timers must be
+   * cancelled from `preRestart` and `postStop`.
+   *
+   * `TimerScheduler` is not thread-safe, i.e. it must only be used within
+   * the actor that owns it.
+   */
+  def createTimerScheduler(context: AbstractActor.ActorContext): TimerScheduler =
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/Timers.scala]: remove this, it's not useful without the intercept of TimerMessage

**Comment by patriknw** [comment]
I should probably take the mima change here into a separate PR so that we can get mima working in master.

**Comment by patriknw** [comment]
I have rebased this and it's ready, please review.

File: akka-docs/src/main/paradox/scala/scheduler.md [hunk 5]
```diff
@@ -10,6 +10,10 @@ You can schedule sending of messages to actors and execution of tasks
 (functions or Runnable).  You will get a `Cancellable` back that you can call
 `cancel` on to cancel the execution of the scheduled operation.
 
+When scheduling periodic or single messages in an actor to itself it is recommended to
+use the @ref:[Actor Timers](actors.md#actors-timers) instead of using the ``Scheduler``
```
Comment (ktoso) [inline akka-docs/src/main/paradox/scala/scheduler.md]: hehe good ol restructured text style snuck in here, single ` is fine

**Review by ktoso — APPROVED** [review]
LGTM in general :) Happy to see this included :)

File: akka-actor/src/main/scala/akka/actor/Timers.scala [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+   * the previous is cancelled and it's guaranteed that a message from the
+   * previous timer is not received, even though it might already be enqueued
+   * in the mailbox when the new timer is started.
+   */
+  def startSingleTimer(key: Any, msg: Any, timeout: FiniteDuration): Unit
+
+  /**
+   * Check if a timer with a given `key` is active.
+   */
+  def isTimerActive(key: Any): Boolean
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/Timers.scala]: what would be the meaning of such type? would it be the scheduled message type? I think it's just an identifier. We could consider if it makes sense for Typed, but for untyped I think it just makes it complicated.

Comment (ktoso) [inline akka-actor/src/main/scala/akka/actor/Timers.scala]: I was thinking to have the `key` typed, to not accidentally use some wrong key when getting the timer... Not sure if very worth it

Comment (ktoso) [inline akka-actor/src/main/scala/akka/actor/Timers.scala]: Just that the Any feels a bit weird, ok to keep as is, but something to think about hm

File: akka-actor/src/main/scala/akka/actor/dungeon/TimerSchedulerImpl.scala [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] class TimerSchedulerImpl(ctx: ActorContext) extends TimerScheduler {
+  import TimerSchedulerImpl._
+
+  private val log = Logging(ctx.system, classOf[TimerScheduler])
+  private var timers: Map[Any, Timer] = Map.empty
+  private val timerGen = Iterator from 1
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/dungeon/TimerSchedulerImpl.scala]: yes

**Comment by patriknw** [comment]
Incorporated feedback from @ktoso 
One more reviewer, please. @raboof ?

File: akka-actor/src/main/scala/akka/actor/Timers.scala [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+  /**
+   * Start a periodic timer that will send `msg` to the `self` actor at
+   * a fixed `interval`.
+   *
+   * Each timer has a key and if a new timer with same key is started
+   * the previous is cancelled and it's guaranteed that a message from the
+   * previous timer is not received, even though it might already be enqueued
+   * in the mailbox when the new timer is started.
+   */
+  def startPeriodicTimer(key: Any, msg: Any, interval: FiniteDuration): Unit
```
Comment (raboof) [inline akka-actor/src/main/scala/akka/actor/Timers.scala]: Should we have both a `timeout`/`delay` and an `interval` like on `scheduler`?

File: akka-actor-tests/src/test/scala/akka/actor/TimerSpec.scala [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+}
+
+class FsmTimerSpec extends AbstractTimerSpec {
+  override def testName: String = "FSM Timers"
+  override def target(monitor: ActorRef, interval: FiniteDuration, repeat: Boolean, initial: () ⇒ Int = () ⇒ 1): Props =
+    TimerSpec.fsmTarget(monitor, interval, repeat, initial)
+}
+
+abstract class AbstractTimerSpec extends AkkaSpec("""
+  akka.loglevel = DEBUG
```
Comment (raboof) [inline akka-actor-tests/src/test/scala/akka/actor/TimerSpec.scala]: Is this (still) needed?

File: akka-actor/src/main/scala/akka/actor/dungeon/TimerSchedulerImpl.scala [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+    timers.get(key) match {
+      case Some(t) ⇒ cancelTimer(t)
+      case None    ⇒
+    }
+    val nextGen = nextTimerGen()
+
+    val timerMsg = TimerMsg(key, nextGen, this)
+    val task =
+      if (repeat)
+        ctx.system.scheduler.schedule(timeout, timeout, ctx.self, timerMsg)(ctx.dispatcher)
```
Comment (raboof) [inline akka-actor/src/main/scala/akka/actor/dungeon/TimerSchedulerImpl.scala]: Would be nice to also be able to schedule a recurring message without delay

File: akka-docs/src/main/paradox/scala/actors.md [hunk 11]
```diff
@@ -905,6 +905,29 @@ Messages marked with `NotInfluenceReceiveTimeout` will not reset the timer. This
 `ReceiveTimeout` should be fired by external inactivity but not influenced by internal activity,
 e.g. scheduled tick messages.
 
+<a id="actors-timers"></a>
```
Comment (raboof) [inline akka-docs/src/main/paradox/scala/actors.md]: Indeed there should be automatic anchors for all headings (though we don't currently check for dangling references, https://github.com/lightbend/paradox/issues/120

File: akka-docs/src/main/paradox/scala/howto.md [hunk 12]
```diff
@@ -122,47 +122,7 @@ The pattern is described [Discovering Message Flows in Actor System with the Spi
...
-@@@
-
-Scala
-:  @@snip [SchedulerPatternSpec.scala]($code$/scala/docs/pattern/SchedulerPatternSpec.scala) { #schedule-receive }
-
-Java
-:  @@snip [SchedulerPatternTest.java]($code$/java/jdocs/pattern/SchedulerPatternTest.java) { #schedule-receive }
-
-@@@ div { .group-java }
+See @ref:[Actor Timers](actors.md#actors-timers)
```
Comment (raboof) [inline akka-docs/src/main/paradox/scala/howto.md]: Perhaps remove the section here entirely?

File: akka-docs/src/main/paradox/scala/scheduler.md [hunk 13]
```diff
@@ -10,6 +10,10 @@ You can schedule sending of messages to actors and execution of tasks
 (functions or Runnable).  You will get a `Cancellable` back that you can call
 `cancel` on to cancel the execution of the scheduled operation.
 
+When scheduling periodic or single messages in an actor to itself it is recommended to
+use the @ref:[Actor Timers](actors.md#actors-timers) instead of using the `Scheduler`
+directly.
```
Comment (raboof) [inline akka-docs/src/main/paradox/scala/scheduler.md]: I like how you don't have to go to `context.system.scheduler` for this common case anymore :+1:

File: akka-docs/src/main/paradox/scala/actors.md [hunk 14]
```diff
@@ -905,6 +905,29 @@ Messages marked with `NotInfluenceReceiveTimeout` will not reset the timer. This
...
+
+## Timers, scheduled messages
```
Comment (raboof) [inline akka-docs/src/main/paradox/scala/actors.md]: actors.md is getting quite long, perhaps we should move some things to their own file... not sure what though :).

**Review by raboof — COMMENTED** [review]
Generally LGTM, but consider if we also want a `delay` parameter on `startPeriodicTimer`. Apart from that good change, reasonable test, neat docs!

File: akka-actor/src/main/scala/akka/actor/Timers.scala (hunk 8 above)
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/Timers.scala]: that is very rarely used, as far as I know, and with this it's easy to first schedule a single timer and then the period after that, if different initial delay is needed

File: akka-actor-tests/src/test/scala/akka/actor/TimerSpec.scala (hunk 9 above)
Comment (patriknw) [inline akka-actor-tests/src/test/scala/akka/actor/TimerSpec.scala]: no, I'll remove

File: akka-actor/src/main/scala/akka/actor/dungeon/TimerSchedulerImpl.scala (hunk 10 above)
Comment (patriknw) [inline akka-actor/src/main/scala/akka/actor/dungeon/TimerSchedulerImpl.scala]: without delay, you mean sending the first message immediately?

File: akka-docs/src/main/paradox/scala/actors.md (hunk 14 above)
Comment (patriknw) [inline akka-docs/src/main/paradox/scala/actors.md]: yes, lot's of that restructuring would be nice, but that is for another time

File: akka-docs/src/main/paradox/scala/howto.md (hunk 12 above)
Comment (patriknw) [inline akka-docs/src/main/paradox/scala/howto.md]: I kept it because I think we have linked to it rather much.

**Comment by patriknw** [comment]
rebased and removed DEBUG, will merge after another pr validation

File: akka-actor/src/main/scala/akka/actor/Timers.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (dwijnand) [reply akka-actor/src/main/scala/akka/actor/Timers.scala]: https://github.com/akka/akka/pull/25865

File: akka-actor/src/main/scala/akka/actor/dungeon/TimerSchedulerImpl.scala [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+    }
+    timers = Map.empty
+  }
+
+  def interceptTimerMsg(timerMsg: TimerMsg): OptionVal[AnyRef] = {
+    timers.get(timerMsg.key) match {
+      case None ⇒
+        // it was from canceled timer that was already enqueued in mailbox
+        log.debug("Received timer [{}] that has been removed, discarding", timerMsg.key)
+        OptionVal.None // message should be ignored
```
[the remark this answers is not in the document]
Comment (dwijnand) [reply akka-actor/src/main/scala/akka/actor/dungeon/TimerSchedulerImpl.scala]: Need to forward-port this back-port, back into Akka Typed that's still using `null` like a wild animal.

**Pull request merged**